### DISCUSSION TITLE
Unable to open image, no images defined.

### DISCUSSION THREAD

**Discussion by maiirisu** [ImageMagick/ImageMagick#5619] [Help]
Hi, I'm running imagemagick in WSL 2. I haven't changed anything in the code or directory, but now that I run it this error happens:

`convert-im6.q16: unable to open image './tmp/xslice00000000.ppm': No such file or directory @ error/blob.c/OpenBlob/2874.
convert-im6.q16: no images defined '/tmp/xslice00000000.gif' @ error/convert.c/ConvertImageCommand/3258.
/bin/rm: cannot remove './tmp/xslice00000000.ppm': No such file or directory`

In the warning note generated it simply says:
Error in using ImageMagick convert command.
Error in removing temporary ppm file.

I've copied and pasted files from the output folder in ubuntu to the windows side and I'm not sure if it's introduced a complication. What could I do to solve this? Thank you!

I've tried searching for the solutions in other forums, and some run into this from PDF conversions and could be fixed via policy change, but that's not the case for me.

**Comment by snibgo** [comment]
The first error is:
```
unable to open image './tmp/xslice00000000.ppm': No such file or directory @ error/blob.c/OpenBlob/2874.
```
That generally means exactly what it says: either the directory or file doesn't exist. Perhaps the error could also arise if the file exists but you don't have permission to read it.

And then "rm" gives the same error.

So, what should have created the file?

**Reply by maiirisu** [reply]
It should be the one creating the file. That's the error that pops up when the code gets to where the ppm and gif-creating functions are. I've also checked the permission within the folder, and again, I've been able to create images within the same directory for quite some time too, so I'm not quite sure if I've messed with the folder or if this is something else.

Other codes which also use imagemagick with the same gif and ppm-creating functions work perfectly well.

**Comment by snibgo** [comment]
What version of IM? What are the commands that create and read the PPM file? Does the directory "./tmp" exist?

**Reply by maiirisu** [reply]
My version is : ImageMagick 6.9.10-23 Q16 x86_64 20190101
And yes, the directory ./tmp exists, a folder within the same folder where the code is.

The template for the main code reads like this:
`template <typename T>
 void ImageWriter<T>::writePpm(
     std::string const &fName, ScalarField2D<T> &field, T minVal, T maxVal) const
 {
     writePpmImplementation(fName, field, minVal, maxVal);
 }`

where the writePpmImplementation is:

`template <typename T>
 void ImageWriter<T>::writePpmImplementation(
     std::string const &fName, ScalarField2D<T> &localField, T minVal, T maxVal) const
 {
     if (global::mpi().isMainProcessor()) {
         if (util::fpequal(minVal, maxVal)) {
             minVal = computeMin(localField);
             maxVal = computeMax(localField);
         }
         std::string fullName = global::directories().getImageOutDir() + fName + ".ppm";
         std::ofstream fout(fullName.c_str());
         fout << "P3\n";
         fout << localField.getNx() << " " << localField.getNy() << "\n";
         fout << (colorRange - 1) << "\n";`
  
  `       for (plint iY = localField.getNy() - 1; iY >= 0; --iY) {
             for (plint iX = 0; iX < localField.getNx(); ++iX) {
                 double outputValue = 0.;
                 if (!(minVal == maxVal)) {
                     outputValue =
                         ((double)(localField.get(iX, iY) - minVal) / (double)(maxVal - minVal)
                          * (double)(numColors - 1) / (double)numColors);
                 }
                 if (outputValue < 0.)
                     outputValue = 0.;
                 if (outputValue >= 1.)
                     outputValue = (double)(numColors - 1) / (double)numColors;
                 rgb color = colorMap.get(outputValue);
                 fout << (int)(color.r * (colorRange - 1)) << " "
                      << (int)(color.g * (colorRange - 1)) << " "
                      << (int)(color.b * (colorRange - 1)) << "\n";
             }
         }
     }
 }`

For context, it is done for the velocity values over a lattice. This is from Palabos, an open-source program for Lattice Boltzmann Fluid Simulations.

Within the code I'm running, it appears as such:

`void writePPM(OffLatticeBoundaryCondition3D<T,DESCRIPTOR,Velocity>& bc, plint iT)
{
    Box3D xSlice(param.cxLB, param.cxLB, 0,          param.ny-1, 0,          param.nz-1);
    Box3D ySlice(0,          param.nx-1, param.cyLB, param.cyLB, 0,          param.nz-1);
    Box3D zSlice(0,          param.nx-1, 0,          param.ny-1, param.czLB, param.czLB);

    ImageWriter<T> writer("leeloo");
    writer.writeScaledPpm(createFileName("vnorm_xslice", iT, PADDING), *bc.computeVelocityNorm(xSlice));
    writer.writeScaledPpm(createFileName("vnorm_yslice", iT, PADDING), *bc.computeVelocityNorm(ySlice));
    writer.writeScaledPpm(createFileName("vnorm_zslice", iT, PADDING), *bc.computeVelocityNorm(zSlice));

 }`

Where it takes slices of the computed norm of the velocity field which it prints into a ppm file. 

As for the copying and pasting I've been doing (which may have harmed the folder), I've been transferring STL files in and out of the folder since I'm working with varying the geometry for the code.

**Comment by snibgo** [comment]
So, you have a program that writes a PPM file. Do you check that the program has succeeded, and the PPM file has been created?

What is your "convert" command line?
